omp/solver/batch_cg_kernels.cpp

```diff
@@ -0,0 +1,110 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "core/solver/batch_cg_kernels.hpp"
+
+
+#include "core/solver/batch_dispatch.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The batch Cg solver namespace.
+ *
+ * @ingroup batch_cg
+ */
+namespace batch_cg {
+
+
+namespace {
+
+
+constexpr int max_num_rhs = 1;
+
+
+#include "reference/base/batch_multi_vector_kernels.hpp.inc"
+#include "reference/matrix/batch_csr_kernels.hpp.inc"
+#include "reference/matrix/batch_dense_kernels.hpp.inc"
+#include "reference/matrix/batch_ell_kernels.hpp.inc"
+#include "reference/solver/batch_cg_kernels.hpp.inc"
+
+
+}  // unnamed namespace
+
+
+template <typename T>
+using settings = gko::kernels::batch_cg::settings<T>;
+
+
+template <typename ValueType>
+class kernel_caller {
+public:
+    kernel_caller(std::shared_ptr<const DefaultExecutor> exec,
+                  const settings<remove_complex<ValueType>> settings)
+        : exec_{std::move(exec)}, settings_{settings}
+    {}
+
+    template <typename BatchMatrixType, typename PrecondType, typename StopType,
+              typename LogType>
+    void call_kernel(
+        const LogType& logger, const BatchMatrixType& mat, PrecondType precond,
+        const gko::batch::multi_vector::uniform_batch<const ValueType>& b,
+        const gko::batch::multi_vector::uniform_batch<ValueType>& x) const
+    {
+        using real_type = typename gko::remove_complex<ValueType>;
+        const size_type num_batch_items = mat.num_batch_items;
+        const auto num_rows = mat.num_rows;
+        const auto num_rhs = b.num_rhs;
+        if (num_rhs > max_num_rhs) {
+            GKO_NOT_IMPLEMENTED;
+        }
+
+        const int local_size_bytes =
+            gko::kernels::batch_cg::local_memory_requirement<ValueType>(
+                num_rows, num_rhs) +
+            PrecondType::dynamic_work_size(num_rows,
+                                           mat.get_single_item_num_nnz()) *
+                sizeof(ValueType);
+
+#pragma omp parallel for
+        for (size_type batch_id = 0; batch_id < num_batch_items; batch_id++) {
+            // TODO: Align to cache line boundary
+            // TODO: Allocate and free once per thread rather than once per
+            // work-item.
+            auto local_space = array<unsigned char>(exec_, local_size_bytes);
```

Comment:
That is quite trivial to do. Just put it outside the loop and add `firstprivate(local_space)` to the pragma

Reply:
Yes, that is true, but I believe one of the OMP versions in our container does not support it and ends up giving wrong results.

Reply:
I guess we have a non-conforming OMP implementation? We should probably fix that.
Anyway, a `std::vector` of arrays would also work.

Reply:
Yes, the nvhpc compiler OMP implementations are annoying to deal with, and I believe follow an older OMP standard ?

Reply:
from https://docs.nvidia.com/hpc-sdk/archive/22.7/compilers/hpc-compilers-user-guide/index.html#openmp-use they say that they support the `firstprivate` key word. Also, the failed test with nvhpc isn't about omp.